dashboard/plugins/s3_usage/s3_stats.py

```diff
@@ -0,0 +1,193 @@
+        s3 = boto3.client('s3', 
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key)
+        self.s3_aggregator = S3Aggregator(s3)
+        self.buckets = [bucket['Name'] for bucket in s3.list_buckets()['Buckets'] 
+                                            if re.match(buckets_regexp, bucket['Name'])]
+        self.logger = logger
+        self.ttl = ttl
+
+    def needs_refresh(self):
+        try:
+            return time.time() - os.stat(self.STATS_DB_PATH).st_mtime > self.ttl
+        except FileNotFoundError:
+            return True
+
+    def run(self):
+        with open(self.LOCK_PATH, 'w') as lock:
+            self.logger.debug("Trying to acquire the S3Stats lock")
+            fcntl.flock(lock, fcntl.LOCK_EX)
+            self.logger.debug("S3Stats Lock acquired")
+            if not self.needs_refresh():
+                self.logger.info("No need to refresh S3Stats, exiting")
+                return
+
+            self.sqlite = sqlite3.connect(self.STATS_DB_TMP_PATH, isolation_level="EXCLUSIVE")
+            self.sqlite.execute(self.CREATE_STMT)
+            self.sqlite.commit()
+            self.logger.info(f"Downloading stats for buckets {self.buckets}")
+
+            global_stats = KeyPrefix()
+            for bucket in self.buckets:
+                data = self.load_s3_bucket(bucket)
+                bucket_stats = self.dump_s3_bucket(data, bucket)
+                global_stats += bucket_stats
+
+            self.sqlite.execute(self.INSERT_STMT, ("", global_stats.files, global_stats.size_standard, global_stats.size_ia, global_stats.size_glacier, 0, 0))
+            self.sqlite.commit()
+            self.sqlite.close()
+            os.rename(self.STATS_DB_TMP_PATH, self.STATS_DB_PATH)
+
+    def insert_iterator(self, bucket, data):
+        for key in data:
+            key_fixed = key.rstrip('/')
+            is_leaf = sum([1 for path in data.keys() if path.startswith(key)]) == 1
+            yield f"/{bucket}{key_fixed}", data[key].files, \
+                data[key].size_standard, data[key].size_ia, data[key].size_glacier, \
+                len(key_fixed.split('/')), int(is_leaf)
+
+    def load_s3_bucket(self, bucket):
+        self.logger.info(f"Dumping bucket {bucket}")
+        start = time.time()
+        data = self.s3_aggregator.load(bucket)
+        self.logger.info(f"It took {format(time.time() - start, '.2f')}s to download information about {len(data)} directories "
+                         f"({data['/'].files} files) in bucket {bucket}")
+        return data
+    
+    def dump_s3_bucket(self, data, bucket):
+        self.sqlite.executemany(self.INSERT_STMT, self.insert_iterator(bucket, data))
+        self.sqlite.commit()
+        return data['/']
+
+@dataclass
+class KeyPrefix:
+    files:int = 0
+    size_standard:int = 0
+    size_ia:int = 0
+    size_glacier:int = 0
+
+    def add_file(self, size, storage_class):
```

Comment:
Is there any cost to adding methods to `dataclass`?

Reply:
Nope, it behaves as a standard class, so there is no additional cost here.